### PR TITLE
add docs for custom types and update auto-configuration

### LINKED ISSUE #956: Update Datastore docs type conversions section

### LINKED ISSUE #971: Document @Unindexed in refdoc for Datastore
Follow-up to #965.

### PR DESCRIPTION
**Pull request by dmitry-s**
Fixes #971 , fixes #956

### CODE REVIEW

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 1]
```diff
@@ -218,25 +218,162 @@ This feature is not supported yet.
...
-==== Custom types
+. The user defined type needs to be mapped to one the basic types supported by Cloud Datastore.
 
-This feature is not supported yet.
+. An instance of both Converters (read and write) needs to be passed to the DatastoreCustomConversions constructor,
+which then has to be made available as a `@Bean` for `SpannerEntityProcessor`.
+
+For example:
+
+We would like to have a field of type  `Album` on our `Singer` POJO:
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: ...and store it in a STRING column in Datastore.

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 2]
```diff
@@ -218,25 +218,162 @@ This feature is not supported yet.
...
+				}
+			};
+
+	//Converters to read custom Album type
+	static final Converter<String, Album> STRING_ALBUM_CONVERTER =
+			new Converter<String, Album>() {
+				@Override
+				public Album convert(String s) {
+					String[] parts = s.split(" ");
+					return new Album(parts[0], LocalDate.parse(parts[1], DateTimeFormatter.ISO_DATE));
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: What if there are spaces in the album name? (very likely in real life)

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 3]
```diff
@@ -218,25 +218,162 @@ This feature is not supported yet.
...
+				Arrays.asList(
+						ALBUM_STRING_CONVERTER,
+						STRING_ALBUM_CONVERTER));
+	}
+}
+----
+
+==== Collections and arrays
+
+Arrays and collections (types that implement java.util.Collection) of supported types are supported.
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: java.util.Collection -> `java.util.Collection`

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 4]
```diff
@@ -218,25 +218,162 @@ This feature is not supported yet.
...
+Elements a converted to Cloud Datastore supported types individually. `byte[]` is an exception, it is converted to
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: s/are/of

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 5]
```diff
@@ -298,7 +435,20 @@ the first two reads utilize keys, while the third is executed using a query base
 
 ===== Indexes
 
-This feature is not supported yet.
+By default, all fields are indexed. To disable indexing on a particular field, @Unindexed annotation can be used.
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: markup `@Unindexed` as code.

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-sample/src/main/java/com/example/Singer.java [hunk 6]
```diff
@@ -36,37 +38,16 @@
 	@Field(name = "last_name")
 	private String lastName;
 
+	private ImmutableSet<Album> albums;
```
Comment (meltsufin) [inline spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-sample/src/main/java/com/example/Singer.java]: Add your name on this class as an author.

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 7]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
 
-==== Lists
+==== Custom types
 
-This feature is not supported yet.
+Custom converters can be used extending the type support for user defined types.
 
-==== Custom types
+. Converters need to implement the `org.springframework.core.convert.converter.Converter` interface
+both directions.
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: *in* both directions

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 8]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+
+. The user defined type needs to be mapped to one the basic types supported by Cloud Datastore.
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: one *of* the basic

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 9]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+
+. An instance of both Converters (read and write) needs to be passed to the DatastoreCustomConversions constructor,
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: Markup `DatastoreCustomConversions` as code.

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 10]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+
+Arrays and collections (types that implement `java.util.Collection`) of supported types are supported.
+They are stored as `com.google.cloud.datastore.ListValue`.
+Elements are converted to Cloud Datastore supported types individually. `byte[]` is an exception, it is converted to
+`com.google.cloud.datastore.Blob`.
+
+
+==== Custom Converter for collections
+Users can provide converters from  `List<?>` to the custom collection type.
+Only read converter is necessary, the Collection API is used on write side to convert a collection to
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: on *the* write side

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-sample/src/main/java/com/example/Singer.java (hunk 6 above)
Comment (meltsufin) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-sample/src/main/java/com/example/Singer.java]: Ping.

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 11]
```diff
@@ -298,7 +444,20 @@ the first two reads utilize keys, while the third is executed using a query base
...
 
-This feature is not supported yet.
+By default, all fields are indexed. To disable indexing on a particular field, `@Unindexed` annotation can be used.
+
+Example:
+[source, java]
+----
+import org.springframework.cloud.gcp.data.datastore.core.mapping.Unindexed;
+
+public class ExapleItem {
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: "ExapleItem" misspelled ?

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 12]
```diff
@@ -298,7 +444,20 @@ the first two reads utilize keys, while the third is executed using a query base
...
+	private long indexedField;
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: not a big deal, but this example class doesn't have getters and setters, so it looks strange that its properties are both "private"

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 13]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+For example:
+
+We would like to have a field of type  `Album` on our `Singer` POJO and want it to be stored as a string property:
+[source, java]
+----
+@Entity
+public class Singer {
+
+	@Id
+	private String singerId;
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: [same comment about the private properties and no getters and setters]

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 14]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+	private Album album;
+}
+----
+
+Where Album is a simple class:
+
+[source, java]
+----
+public class Album {
+	final private String albumName;
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: [same comment about the private properties and no getters and setters]

File: spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc [hunk 15]
```diff
@@ -218,25 +218,171 @@ This feature is not supported yet.
...
+Instead of a field of type `Album`,
+we would like to have a field of type `ImmutableSet<Album>`:
+
+[source, java]
+----
+@Entity
+public class Singer {
+
+	@Id
+	private String singerId;
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-docs/src/main/asciidoc/datastore.adoc]: [same comment about the private properties and no getters and setters]

**Pull request merged**